--- ./project-config/boost-1.45.0/project-config.jam	2011-02-24 17:48:32.000000000 +0800
+++ ./boost-1.45.0/project-config.jam	2011-02-25 10:49:44.000000000 +0800
@@ -9,7 +9,7 @@
 # file.
 if ! gcc in [ feature.values <toolset> ]
 {
-    using gcc ; 
+    using gcc : : mips-linux-gnu-g++ : <compileflags>"-EL -mips32 -msoft-float -fPIC" <linkflags>"-EL -mips32 -msoft-float" ; 
 }
 
 project : default-build <toolset>gcc ;
@@ -25,7 +25,3 @@
 
 # These settings are equivivalent to corresponding command-line
 # options.
-option.set prefix : /usr/local ;
-option.set exec-prefix : /usr/local ;
-option.set libdir : /usr/local/lib ;
-option.set includedir : /usr/local/include ;
