diff -Nura ./gstreamer-0.10.35/ltmain.sh ./gstreamer-0.10.35-new/ltmain.sh
--- ./gstreamer-0.10.35/ltmain.sh	2011-06-15 21:13:39.000000000 +0800
+++ ./gstreamer-0.10.35-new/ltmain.sh	2012-11-06 10:47:45.000000000 +0800
@@ -9267,7 +9267,8 @@
 	      *.la)
 		func_basename "$deplib"
 		name="$func_basename_result"
-		eval libdir=`${SED} -n -e 's/^libdir=\(.*\)$/\1/p' $deplib`
+		func_resolve_sysroot "$deplib"
+		eval libdir=`${SED} -n -e 's/^libdir=\(.*\)$/\1/p' $func_resolve_sysroot_result`
 		test -z "$libdir" && \
 		  func_fatal_error "\`$deplib' is not a valid libtool archive"
 		func_append newdependency_libs " ${lt_sysroot:+=}$libdir/$name"
